src/Microsoft.SqlTools.ServiceLayer/QueryExecution/DataStorage/ServiceBufferFileStreamFactory.cs

```diff
@@ -40,18 +42,28 @@ public string CreateFile()
         /// <returns>A <see cref="ServiceBufferFileStreamReader"/></returns>
         public IFileStreamReader GetReader(string fileName)
         {
-            return new ServiceBufferFileStreamReader(new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite), QueryExecutionSettings);
+            return new ServiceBufferFileStreamReader(
+                new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite),
+                QueryExecutionSettings
+            );
         }
 
         /// <summary>
         /// Creates a new <see cref="ServiceBufferFileStreamWriter"/> for writing values out to an
         /// SSMS formatted buffer file, file share is ReadWrite to allow concurrent reads/writes to the file.
         /// </summary>
         /// <param name="fileName">The file to write values to</param>
+        /// <param name="columns">
+        /// The entire list of columns for the result set. They will be filtered down as per the
+        /// request params.
+        /// </param>
         /// <returns>A <see cref="ServiceBufferFileStreamWriter"/></returns>
-        public IFileStreamWriter GetWriter(string fileName)
+        public IFileStreamWriter GetWriter(string fileName, IReadOnlyList<DbColumnWrapper> columns)
```

Comment:
Intentional. In order to output headers in the constructor for SaveAs CSV (the only one that was updated in this PR), we need columns in the constructor. In order to get columns to the constructor, we need columns in the `GetWriter` method. Since all the file writers (CSV, JSON, ServiceBuffer, etc) implement the same `IFileStreamFactory`, even the `ServiceBufferFileStreamFactory` has to take columns on `GetWriter`. It doesn't make any sense for it to since we don't know the columns until we've read the results, so we just pass in `null`. Avoiding this situation was what the bigger refactoring I did in the original PR was all about - split the factories into one that creates ServiceBuffer reader/writer, one that creates SaveAs writers and eliminates the methods that always throw when used on the wrong class. 

I've updated the comment to indicate this is argument ignored.

Reply:
Note, I still intend to submit the rest of that original PR since these changes really are hacky without it.